Fix duplicate setting

spacecowboy opened the pull request:
Seems like `rotation_delay` has never been an accessible configuration option. It doesn't seem to be in the manual either. Trying to configure the `threshold` option with a valid suffixed value (like `20m`) doesn't work either since `rotation_delay` will try the same value as an integer.

Also added a test to check for duplicates.

spacecowboy committed:
Give log rotation delay setting a unique name so it's accessible

and doesn't override rotation_threshold.

spacecowboy committed:
Add a test to check for duplicate settings

Example output when a duplicate is detected:

    java.lang.AssertionError: 'store.internal_log.rotation_threshold' in store_internal_log_rotation_delay has already been defined in store_internal_log_rotation_threshold

thobe commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -182,7 +182,8 @@
     @Description( "Minimum time (in seconds) after last rotation of the internal log before it may be rotated again." )
-    public static final Setting<Integer> store_internal_log_rotation_delay = setting("store.internal_log.rotation_threshold", INTEGER, "300", min(0), max( Integer.MAX_VALUE ) );
+    public static final Setting<Integer> store_internal_log_rotation_delay = setting("store.internal_log" +
+            ".rotation_delay", INTEGER, "300", min(0), max( Integer.MAX_VALUE ) );
```
Could we avoid line-breaking these strings?

cleishm commented:
While you're at it, fixing the documentation to consistently indicate that the threshold defaults to 250Mb would be nice (http://neo4j.com/docs/stable/configuration-logical-logs.html shows 25Mb).

MishaDemianenko commented:
@spacecowboy can you please fix @cleishm comment

spacecowboy committed:
Fix manual to say correct default settings for logical logs

spacecowboy commented:
@MishaDemianenko @cleishm done

tinwelint commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -182,7 +182,8 @@
     @Description( "Minimum time (in seconds) after last rotation of the internal log before it may be rotated again." )
-    public static final Setting<Integer> store_internal_log_rotation_delay = setting("store.internal_log.rotation_threshold", INTEGER, "300", min(0), max( Integer.MAX_VALUE ) );
+    public static final Setting<Integer> store_internal_log_rotation_delay =
+            setting("store.internal_log.rotation_delay", INTEGER, "300", min(0), max( Integer.MAX_VALUE ) );
```
I think this should be `DURATION` instead if `INTEGER`, so that we can specify f.ex. "10s"

MishaDemianenko replied:
@tinwelint  for some reason in description it says that it only possible to have seconds here, should that be updated to anything then?

tinwelint replied:
I see no reason why this wouldn't be a proper duration. It's also backwards compatible since the default unit is seconds. So please update the description as well.

MishaDemianenko replied:
@tinwelint agree,
@spacecowboy that will also require proper migration of current value from number to duration.

tinwelint replied:
@spacecowboy @MishaDemianenko no migration needed. DURATION is a superset of INTEGER, right?

MishaDemianenko replied:
ah, now i see @tinwelint  thx :)

tinwelint commented:
@spacecowboy @MishaDemianenko no migration needed

spacecowboy committed:
Update log_rotation_delay to be a DURATION

Compatible with old values since default unit is seconds,
and DURATION takes Long values whereas the setting
previously was INTEGERS.

spacecowboy commented:
@tinwelint @MishaDemianenko Agree with your comments. Changed to duration.

MishaDemianenko commented:
@spacecowboy compilation failure

spacecowboy commented:
@MishaDemianenko green now

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -181,8 +181,9 @@
 
-    @Description( "Minimum time (in seconds) after last rotation of the internal log before it may be rotated again." )
-    public static final Setting<Integer> store_internal_log_rotation_delay = setting("store.internal_log.rotation_threshold", INTEGER, "300", min(0), max( Integer.MAX_VALUE ) );
+    @Description( "Minimum time interval after last rotation of the internal log before it may be rotated again." )
```
@spacecowboy  This should produce time in milliseconds, `RotatingFileOutputStreamSupplier` still assume they are seconds  and will try to convert them back to milliseconds by calling `SECONDS.toMillis( rotationDelaySeconds )`

spacecowboy replied:
@MishaDemianenko good catch. Changed it to not convert it. Also the variables were called `second` but actually stored `milliseconds`.

tinwelint replied:
@spacecowboy `RotatingFileOutputStreamSupplier` does the right thing, it's just the `rotationDelaySeconds` field that is badly named.

tinwelint replied:
Oh, I was wrong about this.... DURATION returns `ms` data after all.

tinwelint commented:
To end this ms vs s madness once and for all we should introduce a `Duration` class that is a `long` AND a `TimeUnit` and pass such around instead of a number for time.

spacecowboy commented:
@MishaDemianenko as discussed with @tinwelint , the units are actually correct. IMO we merge this and leave `TimeUnit` for another day. WDYT?

MishaDemianenko commented on community/logging/src/main/java/org/neo4j/logging/RotatingFileOutputStreamSupplier.java:
```diff
@@ -94,38 +92,43 @@ public long getAsLong()
      * @throws IOException If the output file cannot be created
      */
-    public RotatingFileOutputStreamSupplier( FileSystemAbstraction fileSystem, File outputFile, long rotationThresholdBytes, int rotationDelaySeconds, int maxArchives, Executor rotationExecutor ) throws IOException
+    public RotatingFileOutputStreamSupplier( FileSystemAbstraction fileSystem, File outputFile, long rotationThresholdBytes, int rotationDelay, int maxArchives, Executor rotationExecutor ) throws IOException
```
`rotationDelay` should be long instead of int

MishaDemianenko commented:
@spacecowboy one small comment to be fixed
